fix(vm/import): make the UI respond instantly to .ova file drop

nraynaud opened the pull request:
When dropping a file on the import zone, the UI didn't acknowledge the file until the vmdk table were parsed.
Now the UI parses the XML instantly, displays it on the UI, and start parsing the tables in the background. If the user clicks the "import" button, the system will finish parsing the tables and start the upload in when fell swoop. Hiding the parsing time in the upload time.
#### Before: 
![OVA broken UI mp4](https://user-images.githubusercontent.com/46941/93597984-454c0800-f9bc-11ea-898e-2d5d0ff2924c.gif)
#### After:
![fixed OVA UI 1 mp4](https://user-images.githubusercontent.com/46941/93599076-dbccf900-f9bd-11ea-8a97-8797b53db3b0.gif)

No performance improvement has been performed, it's all about UI response, the 12 minutes of before are pushed to after the UI has been updated.

The PR is mostly juggling around some promises.

### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [x] enhancement/bug fix entry added
  - [x] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

nraynaud committed:
VM import: fix UI behavior

When dropping a file on the import zone, the UI didn't acknowledge the file until the vmdk table were parsed.
Now the UI parses the XML instantly, displays it on the UI, and start parsing the tables in the background. If the user clicks the "import" button, the system will finish parsing the tables and start the upload in when fell swoop. Hiding the parsing time in the upload time.

nraynaud committed:
VM import: fix UI behavior

When dropping a file on the import zone, the UI didn't acknowledge the file until the vmdk table were parsed.
Now the UI parses the XML instantly, displays it on the UI, and start parsing the tables in the background. If the user clicks the "import" button, the system will finish parsing the tables and start the upload in when fell swoop. Hiding the parsing time in the upload time.

julien-f commented on packages/xo-vmdk-to-vhd/src/ova.js:
```diff
@@ -342,7 +342,7 @@ export async function parseOVAFile(
         }
       }
-      data.tables[header.fileName] = await readVmdkGrainTable(readFile)
+      data.tables[header.fileName] = readVmdkGrainTable(readFile)
```
A promise that can reject must always have a `catch` listener (or `await`) added before the next async loop.

If that's the case here, you can do this:

```suggestion
      const promise = readVmdkGrainTable(readFile)
      promise.catch(noop) // prevent unhandled rejection
      data.tables[header.fileName] = promise
```

nraynaud replied:
do I really need to do that? I'd be happy for any error to be thrown as an exception when the `await` will block on this promise.

julien-f replied:
That will still be the case: the rejection will always be part of the promise and observable.

But you can't have a rejected promise without a listener at any given time

nraynaud replied:
ok, I will do the fix.

nraynaud replied:
I did it inside `readVmdkGrainTable()`

julien-f replied:
I don't think that's the right place for it, it's too hidden.

Also, it does not work because the `await` in `readVmdkGrainTable` can throw which will reject the returned promise (which does not have a reject listener).

julien-f commented on packages/xo-web/src/xo-app/vm-import/index.js:
```diff
@@ -259,28 +259,32 @@ export default class Import extends Component {
     this.setState({
       vms: [],
     })
+    document.body.style.cursor = 'wait'
```
Wouldn't that be better to save the previous value and reset it at the end?

nraynaud replied:
I removed it altogether, I think it's not really necessary

julien-f commented on packages/xo-vmdk-to-vhd/src/ova.js:
```diff
@@ -342,7 +342,7 @@ export async function parseOVAFile(
         }
       }
-      data.tables[header.fileName] = await readVmdkGrainTable(readFile)
+      data.tables[header.fileName] = readVmdkGrainTable(readFile)
```
Where has this `await` moved to?

Who is using this value?

nraynaud replied:
it's here: 
https://github.com/vatesfr/xen-orchestra/blob/4264e34ffd437daf331bc5790520fec1c9005190/packages/xo-web/src/common/xo/index.js#L1543

nraynaud committed:
VM import: fix UI behavior

When dropping a file on the import zone, the UI didn't acknowledge the file until the vmdk table were parsed.
Now the UI parses the XML instantly, displays it on the UI, and start parsing the tables in the background. If the user clicks the "import" button, the system will finish parsing the tables and start the upload in when fell swoop. Hiding the parsing time in the upload time.

nraynaud committed:
VM import: fix UI behavior

When dropping a file on the import zone, the UI didn't acknowledge the file until the vmdk table were parsed.
Now the UI parses the XML instantly, displays it on the UI, and start parsing the tables in the background. If the user clicks the "import" button, the system will finish parsing the tables and start the upload in when fell swoop. Hiding the parsing time in the upload time.

nraynaud committed:
VM import: fix promise best practices

nraynaud committed:
VM import: try to fix tests

nraynaud committed:
improve unhandled promise rejection risk

julien-f commented on packages/xo-vmdk-to-vhd/src/ova.js:
```diff
@@ -298,6 +298,10 @@ export async function parseOVAFile(
   stringDeserializer,
   skipVmdk = false
 ) {
+  const suppressUnhandledPromise = p => {
```
```suggestion
  const suppressUnhandledRejection = p => {
```

julien-f commented on packages/xo-vmdk-to-vhd/src/vmdk-read-table.js:
```diff
@@ -138,5 +138,8 @@ export async function readCapacityAndGrainTable(fileAccessor) {
-  return { tablePromise: readTable(), capacityBytes: capacity }
+  const tablePromise = readTable()
+  // avoid unhandled promise rejection
+  tablePromise.catch(Function.prototype)
```
Use `suppressUnhandledRejection`.

nraynaud committed:
improve code style

julien-f commented on packages/xo-vmdk-to-vhd/src/ova.js:
```diff
@@ -298,6 +298,10 @@ export async function parseOVAFile(
+  const suppressUnhandledRejection = p => {
+    p.catch(Function.prototype)
+    return p
+  }
```
Mutualize in its own module.

nraynaud committed:
Merge branch 'master' into nr-ova-frozen-ui

# Conflicts:
#	CHANGELOG.unreleased.md

nraynaud committed:
code style

julien-f committed:
Merge branch 'master' into nr-ova-frozen-ui
